File: database/schemas/bus.js
// Bus Schema for MongoDB
// Collection: buses

{
  "_id": ObjectId,
  "operatorId": ObjectId, // Reference to users collection (operator)
  "busNumber": String, // Unique, required, e.g., "MH-01-AB-1234"
  "busName": String, // Required, e.g., "Volvo Multi-Axle"
  "busType": String, // Enum: SLEEPER, SEMI_SLEEPER, SEATER, AC_SLEEPER, NON_AC
  "totalSeats": Number, // Required
  "seatLayout": {
    "type": String, // 2x2, 2x1, etc.
    "rows": Number,
    "columns": Number,
    "seats": [{
      "seatNumber": String, // e.g., "A1", "B2"
      "seatType": String, // WINDOW, AISLE, MIDDLE, UPPER_BERTH, LOWER_BERTH
      "isLadiesSeat": Boolean,
      "isAvailable": Boolean,
      "price": Number, // Dynamic pricing per seat
      "amenities": [String] // LEG_SPACE, CHARGING, WIFI, etc.
    }]
  },
  "amenities": [String], // AC, WIFI, CHARGING, ENTERTAINMENT, BLANKET, PILLOW, WATER, SNACKS
  "features": {
    "isEcoFriendly": Boolean,
    "noiseRating": Number, // 1-5
    "comfortScore": Number, // AI-generated, 1-10
    "hasSilentCabin": Boolean,
    "ladiesSeatZone": Boolean
  },
  "images": [String], // URLs
  "digitalTwin3D": String, // URL to 3D model (optional)
  "registrationNumber": String, // Required
  "insuranceNumber": String,
  "insuranceExpiry": Date,
  "fitnessExpiry": Date,
  "permitExpiry": Date,
  "isActive": Boolean, // Default: true
  "isApproved": Boolean, // Default: false (admin approval required)
  "rating": Number, // Average rating, 1-5
  "totalReviews": Number, // Default: 0
  "createdAt": Date,
  "updatedAt": Date
}

// Indexes
db.buses.createIndex({ "operatorId": 1 });
db.buses.createIndex({ "busNumber": 1 }, { unique: true });
db.buses.createIndex({ "busType": 1 });
db.buses.createIndex({ "isActive": 1, "isApproved": 1 });
db.buses.createIndex({ "rating": -1 });

File: src/main/java/com/busberry/model/Bus.java
package com.busberry.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

/**
 * Bus Model
 * Represents bus fleet information
 */
@Document(collection = "buses")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Bus {

    @Id
    private String id;

    @Indexed
    private String operatorId;

    @Indexed(unique = true)
    private String busNumber;

    private String busName;
    private String busType; // SLEEPER, SEMI_SLEEPER, SEATER, AC_SLEEPER, NON_AC

    private Integer totalSeats;

    private SeatLayout seatLayout;

    private List<String> amenities; // AC, WIFI, CHARGING, etc.

    private BusFeatures features;

    private List<String> images;
    private String digitalTwin3D; // URL to 3D model

    private String registrationNumber;
    private String insuranceNumber;
    private LocalDate insuranceExpiry;
    private LocalDate fitnessExpiry;
    private LocalDate permitExpiry;

    @Builder.Default
    private Boolean isActive = true;
    @Builder.Default
    private Boolean isApproved = false; // Admin approval required

    @Builder.Default
    private Double rating = 0.0;
    @Builder.Default
    private Integer totalReviews = 0;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class SeatLayout {
        private String type; // 2x2, 2x1, etc.
        private Integer rows;
        private Integer columns;
        private List<Seat> seats;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Seat {
        private String seatNumber;
        private String seatType; // WINDOW, AISLE, MIDDLE, UPPER_BERTH, LOWER_BERTH
        private Boolean isLadiesSeat = false;
        private Boolean isAvailable = true;
        private Double price;
        private List<String> amenities;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class BusFeatures {
        @Builder.Default
        private Boolean isEcoFriendly = false;
        @Builder.Default
        private Integer noiseRating = 3; // 1-5
        @Builder.Default
        private Double comfortScore = 5.0; // AI-generated, 1-10
        @Builder.Default
        private Boolean hasSilentCabin = false;
        @Builder.Default
        private Boolean ladiesSeatZone = false;
    }
}

